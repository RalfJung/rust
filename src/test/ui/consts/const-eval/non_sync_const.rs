#![feature(optin_builtin_traits,const_let)]
use std::cell::Cell;

struct Foo;
impl !Sync for Foo {}

struct Bar(i32);
impl !Sync for Bar {}

const FOO1 : &Foo = &Foo; //~ ERROR cannot borrow
const FOO2 : Option<&Foo> = Some(&Foo); //~ ERROR cannot borrow
//~^ ERROR borrowed value does not live long enough
const FOO3 : &Option<Foo> = &Some(Foo); //~ ERROR cannot borrow

const BAR1 : &Bar = &Bar(42); //~ ERROR cannot borrow
const BAR2 : Option<&Bar> = Some(&Bar(42)); //~ ERROR cannot borrow
//~^ ERROR borrowed value does not live long enough
const BAR3 : &Option<Bar> = &Some(Bar(42)); //~ ERROR cannot borrow

/// Tests making sure that value-based reasoning does not go too far.
enum MyOption1<T> { Some(Cell<T>), None } // Never Sync, we shouldnt assume it has Sync values

trait NotSync {} // private trait!
impl NotSync for Foo {}
enum MyOption2<T: NotSync> { Some(T), None } // Never Sync, we shouldnt assume it has Sync values

const OPT1 : &MyOption1<i32> = &MyOption1::None; //~ ERROR cannot borrow
const OPT2 : &MyOption2<Foo> = &MyOption2::None; //~ ERROR cannot borrow

// Hiding in closure variables
const TESTS: &FnOnce() = &{ //~ ERROR cannot borrow
    let x = Cell::new(32);
    move || {
        x;
    }
};

fn main() {}
