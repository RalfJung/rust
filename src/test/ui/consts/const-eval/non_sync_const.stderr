error[E0492]: cannot borrow a constant which may contain interior mutability or non-`Sync` data. If your data is `Sync`, create a static instead
  --> $DIR/non_sync_const.rs:10:21
   |
LL | const FOO1 : &Foo = &Foo; //~ ERROR cannot borrow
   |                     ^^^^

error[E0492]: cannot borrow a constant which may contain interior mutability or non-`Sync` data. If your data is `Sync`, create a static instead
  --> $DIR/non_sync_const.rs:11:34
   |
LL | const FOO2 : Option<&Foo> = Some(&Foo); //~ ERROR cannot borrow
   |                                  ^^^^

error[E0597]: borrowed value does not live long enough
  --> $DIR/non_sync_const.rs:11:35
   |
LL | const FOO2 : Option<&Foo> = Some(&Foo); //~ ERROR cannot borrow
   |                                   ^^^- temporary value only lives until here
   |                                   |
   |                                   temporary value does not live long enough
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0492]: cannot borrow a constant which may contain interior mutability or non-`Sync` data. If your data is `Sync`, create a static instead
  --> $DIR/non_sync_const.rs:13:29
   |
LL | const FOO3 : &Option<Foo> = &Some(Foo); //~ ERROR cannot borrow
   |                             ^^^^^^^^^^

error[E0492]: cannot borrow a constant which may contain interior mutability or non-`Sync` data. If your data is `Sync`, create a static instead
  --> $DIR/non_sync_const.rs:15:21
   |
LL | const BAR1 : &Bar = &Bar(42); //~ ERROR cannot borrow
   |                     ^^^^^^^^

error[E0492]: cannot borrow a constant which may contain interior mutability or non-`Sync` data. If your data is `Sync`, create a static instead
  --> $DIR/non_sync_const.rs:16:34
   |
LL | const BAR2 : Option<&Bar> = Some(&Bar(42)); //~ ERROR cannot borrow
   |                                  ^^^^^^^^

error[E0597]: borrowed value does not live long enough
  --> $DIR/non_sync_const.rs:16:35
   |
LL | const BAR2 : Option<&Bar> = Some(&Bar(42)); //~ ERROR cannot borrow
   |                                   ^^^^^^^- temporary value only lives until here
   |                                   |
   |                                   temporary value does not live long enough
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0492]: cannot borrow a constant which may contain interior mutability or non-`Sync` data. If your data is `Sync`, create a static instead
  --> $DIR/non_sync_const.rs:18:29
   |
LL | const BAR3 : &Option<Bar> = &Some(Bar(42)); //~ ERROR cannot borrow
   |                             ^^^^^^^^^^^^^^

error[E0492]: cannot borrow a constant which may contain interior mutability or non-`Sync` data. If your data is `Sync`, create a static instead
  --> $DIR/non_sync_const.rs:27:32
   |
LL | const OPT1 : &MyOption1<i32> = &MyOption1::None; //~ ERROR cannot borrow
   |                                ^^^^^^^^^^^^^^^^

error[E0492]: cannot borrow a constant which may contain interior mutability or non-`Sync` data. If your data is `Sync`, create a static instead
  --> $DIR/non_sync_const.rs:28:32
   |
LL | const OPT2 : &MyOption2<Foo> = &MyOption2::None; //~ ERROR cannot borrow
   |                                ^^^^^^^^^^^^^^^^

error[E0492]: cannot borrow a constant which may contain interior mutability or non-`Sync` data. If your data is `Sync`, create a static instead
  --> $DIR/non_sync_const.rs:31:26
   |
LL |   const TESTS: &FnOnce() = &{ //~ ERROR cannot borrow
   |  __________________________^
LL | |     let x = Cell::new(32);
LL | |     move || {
LL | |         x;
LL | |     }
LL | | };
   | |_^

error: aborting due to 11 previous errors

Some errors occurred: E0492, E0597.
For more information about an error, try `rustc --explain E0492`.
